src/builtins.c

```diff
@@ -940,22 +940,33 @@ JL_CALLABLE(jl_f_setfield)
     enum jl_memory_order order = jl_memory_order_notatomic;
     JL_NARGS(setfield!, 3, 4);
     if (nargs == 4) {
-        JL_TYPECHK(getfield, symbol, args[3]);
+        JL_TYPECHK(setfield!, symbol, args[3]);
         order = jl_get_atomic_order_checked((jl_sym_t*)args[3], 0, 1);
     }
     jl_value_t *v = args[0];
     jl_datatype_t *st = (jl_datatype_t*)jl_typeof(v);
-    size_t idx = get_checked_fieldindex("setfield!", st, v, args[1], 1);
-    int isatomic = !!jl_field_isatomic(st, idx);
-    if (isatomic == (order == jl_memory_order_notatomic))
-        jl_atomic_error(isatomic ? "setfield!: atomic field cannot be written non-atomically"
-                                 : "setfield!: non-atomic field cannot be written atomically");
-    jl_value_t *ft = jl_field_type_concrete(st, idx);
-    if (!jl_isa(args[2], ft))
-        jl_type_error("setfield!", ft, args[2]);
-    if (order >= jl_memory_order_acq_rel || order == jl_memory_order_release)
-        jl_fence(); // `st->[idx]` will have at least relaxed ordering
-    set_nth_field(st, v, idx, args[2], isatomic);
+    if (st == jl_module_type) {
```

Comment:
In recent discussions, @JeffBezanson considered `getfield` on modules to be a mistake and preferred a separate builtin instead for 2.0. Maybe we should avoid repeating this for setfield! and just go straight for a separate builtin?

Reply:
Maybe, but I think with the current `getfield` that would arguably be more of a special case, no? I'd be happy to change that though.

Reply:
@JeffBezanson should say which he prefers. I had suggested implementing the `getbinding` builtin right away also and just keeping getfield working with the intent to deprecate in 2.0. That way there wouldn't be an asymmetry and `getfield` would just have a bit of a legacy quirk.